torch/_subclasses/fake_tensor.py

```diff
@@ -44,6 +45,41 @@ def _is_tensor_constructor(func: OpOverload):
     )
 
 
+# Similar to `MetaConverter`, this is a class for converting
+# multiple tensors into fake tensors which share the same view/storage
+# structure. Like `MetaConverter`, it will keep alive all
+# tensors that are converted to FakeTensors.
+class FakeTensorConverter(MetaConverter):
+    def __init__(self):
+        self.tensor_memo = {}
+        self.meta_converter = MetaConverter()
+
+    def fake_tensor(self, t, device):
+        if t not in self.tensor_memo:
+            if device:
+                self.tensor_memo[t] = FakeTensor(t, device)
+            else:
+                self.tensor_memo[t] = FakeTensor.from_tensor(t)
+        return self.tensor_memo[t]
+
+    def __call__(self, t, device):
+        return self.fake_tensor(t, device)
+
+
+# We keep one instantiation of `fake_tensor_converter` active
+# for the duration of `with torch_enable_mode(FakeTensorMode)`.
+# This allows accurate storage aliasing across invocation of
+# different operators. While this will keep all freshly allocated
+# tensors alive during `FakeTensorMode`, there will no be no
+# new allocations of Tensors which have non-meta storage so
+# memory should not significantly incraese.
+fake_tensor_converter = None
+
+# if someone invokes `FakeTensorMode` multiple times, we should
+# only clear the fake tensor converter when all modes are done
+active_fake_tensor_modes = 0
```

Comment:
There's an argument for using a stack here... happy to do either

Reply:
This doesn't look compositional. But I know that you were also wondering what it meant to be compositional so let me try to explain.

The primary characteristic of compositionality is that you can reason about various transforms independently and by doing so understand with the whole system does. In other words if there is a fake tensor mode you can understand what it does and it doesn't have any spooky at a distance interactions with further fake tensor modes elsewhere on the stack. The model of compositionality for modes is that any given mode will transform function calls to other function calls, which then get handled by the next mode on the stack. That is to say I could eliminate any given mode simply by rewriting my program to look like the result of having applied that transformation.

With this in mind let's think about what is going on with the code here. You do have action at a distance: multiple instances of the mode share the same converter. You can't just eliminate one of the modes without changing the behavior of the other mode. What I have suggested you do instead is to maintain a separate converter per recursive mode. This way there is no interaction between the modes. I don't really know under what circumstances a recursive mode would be useful here, but I don't really need to know about the use case because the principle of composition tells me what the behavior should be.

Does that help?